### PR TITLE
Implement planpreview handler

### PR DESCRIPTION
**Pull request by nghialv**
**What this PR does / why we need it**:

This PR continues implementing the PlanPreview feature described in this [RFC](https://github.com/pipe-cd/pipe/blob/master/docs/rfcs/0005-plan-preview.md).

In this PR, I have added a new `handler` component that will run when piped started up.
This handler periodically fetches all un-handled BuildPlanPreview commands and split them into a pool of workers.
Each worker handles command serially by making a `builder` instance to build appropriate PlanPreview results for the command.

Which are the main parts that should be reviewed:
- the `handler` struct and its behaviour defined inside `planpreview` package
 
**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by nakabonne** [comment]
I know it's a bit tedious and likely to have no sense, but it would be helpful to review if you could write down a brief summary or where you want us to look over if there is. No need to do it every time, but it's gonna make for a good discussion if you could give it a try as much as possible 😃

File: pkg/app/piped/planpreview/builder.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	apps := b.listApplications(repoCfg)
+	if len(apps) == 0 {
+		return nil, nil
+	}
+
+	repo, err := b.gitClient.Clone(ctx, repoCfg.RepoID, repoCfg.Remote, repoCfg.Branch, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to clone git repository %s", cmd.RepositoryId)
+	}
+	defer repo.Clean()
```
Comment (nakabonne) [inline pkg/app/piped/planpreview/builder.go:65]: I didn't know there is such a helper.

**Comment by nakabonne** [comment]
Cool. there you go!
/lgtm

**Comment by nghialv** [comment]
> I know it's a bit tedious and likely to have no sense, but it would be helpful to review if you could write down a brief summary or where you want us to look over if there is. No need to do it every time, but it's gonna make for a good discussion if you could give it a try as much as possible 😃

Sure. I have just updated the PR body to explain more about this PR.

**Comment by nakabonne** [comment]
Thank you!

File: pkg/app/piped/planpreview/handler.go [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+	builderFactory func() Builder
+	logger         *zap.Logger
+}
+
+func NewHandler(gc gitClient, cl commandLister, al applicationLister, cfg *config.PipedSpec, opts ...Option) *Handler {
+	opt := &options{
+		workerNum:              3,
+		commandQueueBufferSize: 10,
+		commandCheckInterval:   5 * time.Second,
+		commandHandleTimeout:   5 * time.Minute,
```
Comment (nakabonne) [inline pkg/app/piped/planpreview/handler.go:99]: nit: they should be defined as const like `default~`, some other places don't be so though.

Comment (khanhtc1202) [reply pkg/app/piped/planpreview/handler.go:99]: Agree, make them as const variables of this `planpreview` package is LGTM 👍

Comment (nghialv) [reply pkg/app/piped/planpreview/handler.go]: Sure. Applied this. 👍

**Comment by khanhtc1202** [comment]
Nice 🚀 
/approve

**Pull request merged**